### DISCUSSION TITLE
How should we type custom items?

### DISCUSSION THREAD

**Discussion by IanVS** [ariakit/ariakit#2508] [Questions] [answered]
I like to keep as much type safety as I can in my forms, so I typically treat my options as const.  Here's a simplified example:

```ts
const OPTIONS = ['a', 'b', 'c'] as const;
const OPTION_LABELS = {a: 'A', b: 'B', c: 'C'};

function renderValue(val: typeof OPTIONS[number]) {
  return OPTION_LABELS[val];
}

const select = Ariakit.useSelectStore();

// This is typed as `Value`, essentially: string | string[]
const value = select.useState('value');

<Ariakit.Select store={select}>
  {/* This fails, because value is not strictly typed */} 
  {renderValue(value)}
</Ariakit.Select>
<Ariakit.SelectPopover store={select}>
  {OPTIONS.map((opt) => {
    // This part works okay, because opt is strictly typed
    <Ariakit.SelectItem className="select-item" value={opt}>{renderValue(opt)}</Ariakit.SelectItem>
  }}
</Ariakit.SelectPopover>
```

How would you suggest working with custom items like this?  For now I'm using ts-expect-error comments, but I'd love to find a good way to wire this up correctly.  And BTW I tried using a generic to `useSelectStore`, but that didn't work either, because it turns that generic into `MutableValue<T>`, and other type errors start to pop up.

**Comment by diegohaz** [comment]
Can you set a `defaultValue` on `useSelectStore`?

**Reply by IanVS** [reply]
I did try that in various ways.  Here's a stackblitz example of one such attempt: https://stackblitz.com/edit/tlswsk?file=select-item-custom%2Findex.tsx

**Reply by diegohaz** [reply, marked as answer]
I'd cast the value to `T` in that case: https://stackblitz.com/edit/tlswsk-g1a3rc?file=select-item-custom%2Findex.tsx%3AL40

`MutableValue` was added to select so the value type could be inferred from the `defaultValue` type without people having to write specific TS code.

With `MutableValue`:
```ts
const select = useSelectStore({ defaultValue: "" })
const value = select.useState("value") // string
```

Without `MutableValue`:

```ts
const select = useSelectStore({ defaultValue: "" })
const value = select.useState("value") // ""
```

```ts
const select = useSelectStore({ defaultValue: "" as string })
const value = select.useState("value") // string
```

```ts
const select = useSelectStore<string>({ defaultValue: "" })
const value = select.useState("value") // string
```

I don't know if there's a better way to handle this. But I'm open to suggestions.
